fieldpath/set.go

```diff
@@ -136,6 +138,126 @@ func (s *Set) EnsureNamedFieldsAreMembers(sc *schema.Schema, tr schema.TypeRef)
 	}
 }
 
+// MustPrefixPattern is the same as PrefixPattern except it panics if parts can't be
+// turned into a SetPattern.
+func MustPrefixPattern(parts ...interface{}) *SetPattern {
+	result, err := PrefixPattern(parts...)
+	if err != nil {
+		panic(err)
+	}
+	return result
+}
+
+// PrefixPattern creates a SetPattern that matches all field paths prefixed by the given list of pattern path parts.
+// The pattern parts may any of:
+//
+//   - PathPattern - for wildcards, `MatchAnyPathElement()` can be used as well.
+//   - PathElement - for any path element
+//   - value.FieldList - for associative list keys
+//   - value.Value - for scalar list elements
+//   - string - For field names
+//   - int - for array indices
+func PrefixPattern(parts ...interface{}) (*SetPattern, error) {
+	current := MatchAnySet() // match all field patch suffixes
+	for i := len(parts) - 1; i >= 0; i-- {
+		part := parts[i]
+		var pattern PathPattern
+		switch t := part.(type) {
+		case PathPattern:
+			pattern = t
+		case PathElement:
+			pattern = PathPattern{PathElement: t}
+		case *value.FieldList:
+			if len(*t) == 0 {
+				return nil, fmt.Errorf("associative list key type path elements must have at least one key (got zero)")
+			}
+			pattern = PathPattern{PathElement: PathElement{Key: t}}
+		case value.Value:
+			pattern = PathPattern{PathElement: PathElement{Value: &t}}
+		case string:
+			pattern = PathPattern{PathElement: PathElement{FieldName: &t}}
+		case int:
+			pattern = PathPattern{PathElement: PathElement{Index: &t}}
+		default:
+			return nil, fmt.Errorf("unexpected type %T", t)
+		}
+		current = &SetPattern{
+			Members: []*MemberSetPattern{{
+				Path:  pattern,
+				Child: current,
+			}},
+		}
+	}
+	return current, nil
+}
+
+// MatchAnyPathElement returns a PathPattern that matches any path element.
+func MatchAnyPathElement() PathPattern {
+	return PathPattern{Wildcard: true}
+}
+
+// MatchAnySet returns a SetPattern that matches any set.
+func MatchAnySet() *SetPattern {
+	return &SetPattern{Wildcard: true}
+}
+
+// SetPattern defines a pattern that matches fields in a Set.
+// SetPattern is structured much like a Set but with wildcard support.
+type SetPattern struct {
```

Comment:
It's a little awkward being singular here, when it's possibly plural.  This is made a little worse because `NewPatternFilter()` takes a single pattern and `PrefixPattern()` constructs a `SetPattern` with one member.

How is one supposed to construct a pattern with multiple members?  The use-case in question needs multiple, I think.

The recursive use of `SetPattern` is tricky, but maybe allows something like the following:

from the existing test:

```
filter := NewPatternFilter(MustPrefixPattern("spec", "containers", MatchAnyPathElement(), "resources"))
```

could maybe be:

```
filter := NewPatternFilter(
    MustMakePattern("spec", "containers", MatchAnyPathElement(), "resources"),
    MustMakePattern("spec", "initContainers", MatchAnyPathElement(), "resources"),
    MustMakePattern("spec", "resources"),
)
```

And we would combine those into the final `SetPattern` (if len(args) == 1, just use it, else combine them)?

Reply:
A plural name makes sense to me.  How about `SetMatcher`?

```go
filter := NewPatternFilter(
    MustMakePattern("spec", "containers", MatchAnyPathElement(), "resources"),
    MustMakePattern("spec", "initContainers", MatchAnyPathElement(), "resources"),
    MustMakePattern("spec", "resources"),
)
```

This seems like a reasonable way define matcher that branches off in multiple directions.  `SetPattern` does support multiple branches, but hand crafting one as a data literal is non-egronomic to the extreme.

The easiest way to support might be to accept multiple matchers and then union them together.

Reply:
Added merge support and modified `NewIncludeMatcherFilter()` to take matchers as varargs, then merge them together before filtering.

I need someone to double-check me for the  "wildcard takes precedence" test. Is this what we want?